tests/test-helmet.php

```diff
@@ -0,0 +1,317 @@
+		$page_helmet = ( new Component( 'irving/helmet' ) )->set_child( $page_title );
+
+		// Example `meta` component.
+		$meta_example = ( new Component( 'meta' ) )
+			->set_config( 'name', 'robots' )
+			->set_config( 'content', 'noindex, follow' );
+
+		// Example endpoint data.
+		$data = [
+			'defaults'       => [],
+			'page'           => [],
+			'providers'      => [],
+			'redirectTo'     => '',
+			'redirectStatus' => 0,
+		];
+
+		/**
+		 * Test the basic functionality.
+		 */
+		$data_with_helmet = Templates\setup_helmet( $data, new \WP_Query(), 'site', '/', new \WP_Rest_Request() );
+		$this->assertEquals( $default_helmet, $data_with_helmet['defaults'][0] );
+		$this->assertEquals( $page_helmet, $data_with_helmet['page'][0] );
+
+		/**
+		 * Test disabling functionality via filter.
+		 */
+		add_filter( 'wp_irving_setup_helmet', '__return_false' );
+		$data_with_helmet = Templates\setup_helmet( $data, new \WP_Query(), 'site', '/', new \WP_Rest_Request() );
+		$this->assertEquals( [], $data_with_helmet['defaults'] );
+		$this->assertEquals( [], $data_with_helmet['page'] );
+		add_filter( 'wp_irving_setup_helmet', '__return_true' );
+
+		/**
+		 * Test the default helmet filter.
+		 */
+		add_filter(
+			'wp_irving_default_helmet_component',
+			function( $helmet ) use ( $meta_example ) {
+				$helmet->prepend_child( $meta_example ); // Pre-pend so we can test more complex output.
+				return $helmet;
+			}
+		);
+
+		// Run setup with the new helmet filter in place.
+		$data_with_helmet = Templates\setup_helmet( $data, new \WP_Query(), 'site', '/', new \WP_Rest_Request() );
+		$this->assertEquals(
+			[
+				$meta_example,
+				$default_title,
+			],
+			$data_with_helmet['defaults'][0]->get_children()
+		);
+
+		/**
+		 * Test the page helmet filter.
+		 */
+		add_filter(
+			'wp_irving_page_helmet_component',
+			function( $helmet ) use ( $meta_example ) {
+				$helmet->prepend_child( $meta_example ); // Pre-pend so we can test more complex output.
+				return $helmet;
+			}
+		);
+
+		// Run setup with the new helmet filter in place.
+		$data_with_helmet = Templates\setup_helmet( $data, new \WP_Query(), 'site', '/', new \WP_Rest_Request() );
+		$this->assertEquals(
+			[
+				$meta_example,
+				$page_title,
+			],
+			$data_with_helmet['page'][0]->get_children()
+		);
+	}
+
+	/**
+	 * Test the `inject_head_tags` function.
+	 */
+	public function test_inject_head_tags() {
+
+		// Example components.
+		$helmet = new Component( 'irving/helmet' );
+		$title  = ( new Component( 'title' ) )->set_child( 'Irving Development - Just another WordPress site' );
+
+		// Parse the example markup, only grabbing the title tag.
+		$parsed_html = parse_html( $this->example_markup, [ 'title' ] );
+
+		$this->assertEquals(
+			( new Component( 'irving/helmet' ) )->set_child( $title ),
+			inject_head_tags( $helmet, $parsed_html )
+		);
+	}
+
+	/**
+	 * Test the `get_title_from_helmet()` functin.
+	 */
+	public function test_get_title_from_helmet() {
```

Comment:
This is a good candidate for a data provider. When you test several different iterations of test data it's also helpful to pass an extra message parameter to the assertion that describes which test failed, e.g.:

```
$helmet = new Component( 'irving/helmet' );
$this->assertNull( get_title_from_helmet( $helmet ), 'Title not null in default helmet component.' );
```